### PR TITLE
[WIP] Glsk integ

### PR DESCRIPTION
**Pull request by pengbo-ruidian**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
feature glsk

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)
glsk depends on powsybl-core version 2.5.0-snapshot, for latest dev on Scalable.

### CODE REVIEW

**Comment by pengbo-ruidian** [comment]
glsk depends on powsybl-core version 2.5.0-snapshot, for latest dev on Scalable.

File: data/glsk-file/pom.xml [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        <groupId>com.farao-community.farao</groupId>
+        <version>1.1.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>farao-glsk-file</artifactId>
+    <name>GLSK file</name>
+    <description>Module that regroups all features related to GLSK file</description>
+
+    <build>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: This part of the POM file is not needed cause redondant with the one already defined in parent POM. To be removed

File: data/glsk-file/pom.xml [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-api</artifactId>
+            <version>${slf4j.version}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Define the version of the dependencies in the "dependencyMangement" part of the main POM. Then remove the version on the children POMs

File: data/glsk-file/pom.xml [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-action-util</artifactId>
+            <version>${powsybl.core.version-dev}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-commons</artifactId>
+            <version>${powsybl.core.version-dev}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-ucte-network</artifactId>
+            <version>${powsybl.core.version-dev}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-iidm-api</artifactId>
+            <version>${powsybl.core.version-dev}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-iidm-converter-api</artifactId>
+            <version>${powsybl.core.version-dev}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-iidm-xml-converter</artifactId>
+            <version>${powsybl.core.version-dev}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-sensitivity-api</artifactId>
+            <version>${powsybl.core.version-dev}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.farao-community.farao</groupId>
+            <artifactId>farao-chronology</artifactId>
+            <version>0.0.1-SNAPSHOT</version>
+        </dependency>
+        <dependency>
+            <groupId>org.threeten</groupId>
+            <artifactId>threeten-extra</artifactId>
+            <version>${threeten.version}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        </dependency>
+
+
+
+        <!-- Test dependencies -->
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>${junit.version}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-simple</artifactId>
+            <version>${slf4j.version}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-commons</artifactId>
+            <version>${powsybl.core.version-dev}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+            <type>test-jar</type>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-iidm-xml-converter</artifactId>
+            <version>${powsybl.core.version-dev}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/pom.xml [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-iidm-impl</artifactId>
+            <version>${powsybl.core.version-dev}</version>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Same here

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java [hunk 16]
```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Comment (murgeyseb) [inline data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java]: 2019

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.farao_community.farao.commons.data.generatorloadshiftkeys;
```
Comment (murgeyseb) [inline data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java]: Rename package : com.farao_community.farao.data.glsk_file

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+
+import com.powsybl.iidm.network.Country;
+
+/**
+ * EIC = Energy Identification Code
```
Comment (murgeyseb) [inline data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java]: Add @author mention in class header

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java [hunk 19]
```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright (c) 2018, RTE (http://www.rte-france.com)
```
Comment (murgeyseb) [inline data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java]: 2019

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.farao_community.farao.commons.data.generatorloadshiftkeys;
```
Comment (murgeyseb) [inline data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java]: Rename package : com.farao_community.farao.data.glsk_file

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+import javax.xml.parsers.ParserConfigurationException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * CIM type GlskDocument
```
Comment (murgeyseb) [inline data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java]: Add @author mention in class header

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskException.java [hunk 22]
```diff
@@ -0,0 +1,20 @@
+/**
```
Comment (murgeyseb) [inline data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskException.java]: This exception do not give extra information. You may use FaraoException instead ?

File: data/glsk-file/pom.xml [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+            <artifactId>powsybl-iidm-xml-converter</artifactId>
+            <version>${powsybl.core.version-dev}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-sensitivity-api</artifactId>
+            <version>${powsybl.core.version-dev}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.farao-community.farao</groupId>
```
Comment (murgeyseb) [inline data/glsk-file/pom.xml]: Fork the code of *farao-chronology* and put it directly in your merge request as part of farao-core modules (in commons I would say)

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java (hunk 16 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java (hunk 17 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java (hunk 19 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java (hunk 21 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java (hunk 20 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskDocument.java]: ok

File: data/glsk-file/pom.xml (hunk 1 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 11 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 12 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 4 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 6 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 8 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 9 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 13 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 14 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java (hunk 18 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/EICode.java]: ok

File: data/glsk-file/pom.xml (hunk 2 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 5 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 7 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 15 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 10 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml (hunk 3 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskException.java (hunk 22 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/generatorloadshiftkeys/GlskException.java]: ok

File: data/glsk-file/pom.xml (hunk 23 above)
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/pom.xml [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-iidm-xml-converter</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.powsybl</groupId>
+            <artifactId>powsybl-sensitivity-api</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.farao-community.farao</groupId>
+            <artifactId>farao-chronology</artifactId>
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/pom.xml]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/EICode.java [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.farao_community.farao.commons.data.glsk_file;
+
+import com.powsybl.iidm.network.Country;
+
+/**
+ * @author RTE International {@literal <contact@rte-international.com>}
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/EICode.java]: ok

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentLinearGlskConverterTest.java [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Test
+    public void testConvertUcteGlskDocumentToLinearGlskDataChronologyFromFilePathString() throws ParserConfigurationException, SAXException, IOException {
+        String filepathstring = "src/test/resources/20170322_1844_SN3_FR2_GLSK_test.xml";
+        assertTrue(!new UcteGlskDocumentLinearGlskConverter().convertUcteGlskDocumentToLinearGlskDataChronologyFromFilePathString(filepathstring, testNetwork).isEmpty());
+    }
+
+    @Test
+    public void testConvertUcteGlskDocumentToLinearGlskDataChronologyFromFilePath() throws ParserConfigurationException, SAXException, IOException {
+        Path pathtest = Paths.get("src/test/resources/20170322_1844_SN3_FR2_GLSK_test.xml");
```
Comment (murgeyseb) [inline data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentLinearGlskConverterTest.java]: Use Paths.get(getClass().getResource("......").toURI())

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentLinearGlskConverterTest.java [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+        for (String country : mapGlskDocLinearGlsk.keySet()) {
+            DataChronology<LinearGlsk> dataChronology = mapGlskDocLinearGlsk.get(country);
+            assertTrue(dataChronology.getDataForInstant(Instant.parse("2016-07-28T22:00:00Z")).isPresent());
+            assertFalse(dataChronology.getDataForInstant(Instant.parse("2018-08-26T21:00:00Z")).isPresent());
+        }
+    }
+
+    @Test
+    public void testConvertUcteGlskDocumentToLinearGlskDataChronologyFromFilePathString() throws ParserConfigurationException, SAXException, IOException {
+        String filepathstring = "src/test/resources/20170322_1844_SN3_FR2_GLSK_test.xml";
```
Comment (murgeyseb) [inline data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentLinearGlskConverterTest.java]: Use getClass().getResource("......").getPath()

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentScalableConverterTest.java [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Test
+    public void testConvertUcteGlskDocumentToScalableDataChronologyFromFilePathString() throws ParserConfigurationException, SAXException, IOException {
+        String filepathstring = "src/test/resources/20170322_1844_SN3_FR2_GLSK_test.xml";
+        assertTrue(!new UcteGlskDocumentScalableConverter().convertUcteGlskDocumentToScalableDataChronologyFromFilePathString(filepathstring, testNetwork).isEmpty());
+    }
+
+    @Test
+    public void testConvertUcteGlskDocumentToScalableDataChronologyFromFilePath() throws ParserConfigurationException, SAXException, IOException {
+        Path pathtest = Paths.get("src/test/resources/20170322_1844_SN3_FR2_GLSK_test.xml");
```
Comment (murgeyseb) [inline data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentScalableConverterTest.java]: Use Paths.get(getClass().getResource("......").toURI())

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentScalableConverterTest.java [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+        for (String country : mapGlskDocScalable.keySet()) {
+            DataChronology<Scalable> dataChronology = mapGlskDocScalable.get(country);
+            assertTrue(dataChronology.getDataForInstant(Instant.parse("2016-07-28T22:00:00Z")).isPresent());
+            assertFalse(dataChronology.getDataForInstant(Instant.parse("2018-08-26T21:00:00Z")).isPresent());
+        }
+    }
+
+    @Test
+    public void testConvertUcteGlskDocumentToScalableDataChronologyFromFilePathString() throws ParserConfigurationException, SAXException, IOException {
+        String filepathstring = "src/test/resources/20170322_1844_SN3_FR2_GLSK_test.xml";
```
Comment (murgeyseb) [inline data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentScalableConverterTest.java]: Use getClass().getResource("......").getPath()

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/GlskTimeSeries.java [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+    private String curveType;
+    /**
+     * list of periods in the time series
+     */
+    private List<GlskPeriod> glskPeriods;
+
+    /**
+     * @param element Time series element
+     */
+    public GlskTimeSeries(Element element) {
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/GlskTimeSeries.java]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/GlskShiftKey.java [hunk 31]
```diff
@@ -0,0 +1,200 @@
...
+     * merit order direction
+     */
+    private String flowDirection;
+
+    /**
+     * @param element Dom element of CIM Glsk
+     * @param pointInterval interval of point
+     * @param subjectDomainmRID country mrid
+     */
+    public GlskShiftKey(Element element, Interval pointInterval, String subjectDomainmRID) {
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/GlskShiftKey.java]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/GlskPeriod.java [hunk 32]
```diff
@@ -0,0 +1,109 @@
...
+     * List of GlskPoint in the Period
+     */
+    private List<GlskPoint> glskPoints;
+
+    /**
+     * @param element Dom element
+     * @param subjectDomainmRID Country mRID
+     * @param curveType curveType
+     */
+    public GlskPeriod(Element element, String subjectDomainmRID, String curveType) {
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/GlskPeriod.java]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/actors/GlskDocumentImporter.java [hunk 33]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * @param filepath file path in java Path
+     * @return GlskDocument object
+     * @throws IOException
+     * @throws ParserConfigurationException
+     * @throws SAXException
+     */
+    public GlskDocument importGlskDocumentWithFilePath(Path filepath) throws IOException, ParserConfigurationException, SAXException {
+        InputStream data = new FileInputStream(filepath.toFile());
+        return this.importGlskDocumentFromInputStream(data);
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/actors/GlskDocumentImporter.java]: ok

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/GlskDocumentLinearGlskConverterTest.java [hunk 34]
```diff
@@ -0,0 +1,97 @@
...
+        GlskPoint glskPointParticipationFactorGskLsk = importer.importGlskDocumentWithFilename(GLSKB43GSKLSK).getGlskPoints().get(0);
+
+    }
+
+    @Test
+    public void testMappingGlskDocumentToDataChronology() throws ParserConfigurationException, SAXException, IOException {
+        GlskDocumentLinearGlskConverter mapper = new GlskDocumentLinearGlskConverter();
+        Map<String, DataChronology<GlskPoint>> map = mapper.convertGlskDocumentToGlskPointDataChronologyFromFileName("/GlskB42test.xml");
+
+//        for (String country : map.keySet()) {
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/GlskDocumentLinearGlskConverterTest.java]: ok

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/GlskDocumentScalableConverterTest.java [hunk 35]
```diff
@@ -0,0 +1,85 @@
...
+        for (String country : mapGlskDocScalable.keySet()) {
+            DataChronology<Scalable> dataChronology = mapGlskDocScalable.get(country);
+            assertTrue(dataChronology.getDataForInstant(Instant.parse("2018-08-29T21:00:00Z")).isPresent());
+            assertFalse(dataChronology.getDataForInstant(Instant.parse("2018-08-26T21:00:00Z")).isPresent());
+        }
+    }
+
+    @Test
+    public void testConvertGlskDocumentToScalableDataChronologyFromFilePathString() throws ParserConfigurationException, SAXException, IOException {
+        String filepathstring = "src/test/resources/GlskB42CountryIIDM.xml";
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/GlskDocumentScalableConverterTest.java]: ok

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentLinearGlskConverterTest.java (hunk 26 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentLinearGlskConverterTest.java]: ok

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentScalableConverterTest.java (hunk 28 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentScalableConverterTest.java]: ok

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentScalableConverterTest.java (hunk 29 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentScalableConverterTest.java]: ok

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/GlskDocumentScalableConverterTest.java [hunk 36]
```diff
@@ -0,0 +1,85 @@
...
+        assertTrue(!new GlskDocumentScalableConverter().convertGlskDocumentToScalableDataChronologyFromFilePathString(filepathstring, testNetwork).isEmpty());
+    }
+
+    @Test
+    public void testConvertGlskDocumentToScalableDataChronologyFromFilePath() throws ParserConfigurationException, SAXException, IOException {
+        Path pathtest = Paths.get("src/test/resources/GlskB42CountryIIDM.xml");
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/GlskDocumentScalableConverterTest.java]: ok

File: data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentLinearGlskConverterTest.java (hunk 27 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/test/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocumentLinearGlskConverterTest.java]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/actors/GlskDocumentImporter.java [hunk 37]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * @param filepathstring absolute file path in string
+     * @return GLSKDocument object
+     * @throws IOException
+     * @throws ParserConfigurationException
+     * @throws SAXException
+     */
+    public GlskDocument importGlskDocumentWithFilePathString(String filepathstring) throws IOException, ParserConfigurationException, SAXException {
+        InputStream data = new FileInputStream(filepathstring);
+        return this.importGlskDocumentFromInputStream(data);
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/actors/GlskDocumentImporter.java]: ok

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/EICode.java [hunk 38]
```diff
@@ -0,0 +1,55 @@
...
+ */
+package com.farao_community.farao.commons.data.glsk_file;
+
+import com.powsybl.iidm.network.Country;
+
+/**
+ * @author Pengbo Wang {@literal <pengbo.wang@rte-international.com>}
+ * EIC = Energy Identification Code
+ */
+public class EICode {
```
[the remark this answers is not in the document]
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/EICode.java]: I tried to use enum in the beginning but there are special characters in the EICode like: "10YBE----------2", so I find a class is more easy.

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/GlskDocument.java [hunk 39]
```diff
@@ -0,0 +1,93 @@
...
+
+    /**
+     * @param data input file stream
+     * @throws ParserConfigurationException
+     * @throws IOException
+     * @throws SAXException
+     */
+    public GlskDocument(InputStream data) throws ParserConfigurationException, IOException, SAXException {
+        DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+        documentBuilderFactory.setNamespaceAware(true);
```
Comment (murgeyseb) [inline data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/GlskDocument.java]: Add
`documentBuilderFactory.setAttribute(XMLConstants.FEATURE_SECURE_PROCESSING, Boolean.TRUE);
`
It should remove the Sonar error

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocument.java [hunk 40]
```diff
@@ -0,0 +1,156 @@
...
+
+    /**
+     * @param data input file as input stream
+     * @throws ParserConfigurationException
+     * @throws IOException
+     * @throws SAXException
+     */
+    public UcteGlskDocument(InputStream data) throws ParserConfigurationException, IOException, SAXException {
+        DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+        documentBuilderFactory.setNamespaceAware(true);
```
Comment (murgeyseb) [inline data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocument.java]: Add
`documentBuilderFactory.setAttribute(XMLConstants.FEATURE_SECURE_PROCESSING, Boolean.TRUE);
`
It should remove the Sonar error

Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/UcteGlskDocument.java]: Try this solution on commit 34b62e9 but it does not solve the SonarCloud warning. Could be a bug of SonarCloud's rule?

File: data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/GlskDocument.java (hunk 39 above)
Comment (pengbo-ruidian) [reply data/glsk-file/src/main/java/com/farao_community/farao/commons/data/glsk_file/GlskDocument.java]: Try this solution on commit 34b62e9 but it does not solve the SonarCloud warning. Could be a bug of SonarCloud's rule?

**Pull request merged**